scheduler/feasible.go

```diff
@@ -177,6 +185,75 @@ func (h *HostVolumeChecker) hasVolumes(n *structs.Node) bool {
 	return true
 }
 
+type CSIVolumeChecker struct {
+	ctx     Context
+	volumes map[string]*structs.VolumeRequest
+}
+
+func NewCSIVolumeChecker(ctx Context) *CSIVolumeChecker {
+	return &CSIVolumeChecker{
+		ctx: ctx,
+	}
+}
+
+func (c *CSIVolumeChecker) SetVolumes(volumes map[string]*structs.VolumeRequest) {
+	c.volumes = volumes
+}
+
+func (c *CSIVolumeChecker) Feasible(n *structs.Node) bool {
+	if c.hasPlugins(n) {
+		return true
+	}
+
+	c.ctx.Metrics().FilterNode(n, FilterConstraintCSIVolumes)
+	return false
+}
+
+func (c *CSIVolumeChecker) hasPlugins(n *structs.Node) bool {
+	// Have volume if
+	// - healthy controller plugin is running the driver
```

Comment:
We only need a controller plugin if `RequiresControllerPlugin` is true on a target node plugin, otherwise we can skip the controller flow entirely. Afaict we don't handle that here.

We could also provide specific errors around `missing CSI Node Plugin` and `missing CSI Controller Plugin` based on this too.

Reply:
The UX designs have volumes listed with controller health data, so I lifted `RequiresControllerPlugin` up to the volume. The volume health check now accounts for that flag, so there's no change in the feasibility check.

Reply:
That should be fine :+1:

Reply:
Could we update the comment to reflect that actually? - Will be helpful for future readers